### ISSUE TITLE
@BeforeEach and @AfterEach aren't reliably executed around dynamic tests

### ISSUE THREAD

**Issue by ogregoire** [junit-team/junit5#2360]
`@BeforeEach` and `@AfterEach` aren't executed around each dynamic test. They are executed only once per `@TestFactory`. I would expect that they are reliably run around each dynamic test.

I tested on both IntelliJ 2020.1.3 and Maven 3.6.1. What's baffling is that the results are different on each platform.

On IntelliJ, with the snippet below, I get the following output:

    before
    after
    Hello
    World

On Maven, I get the following output:

    before
    Hello
    World
    after

I would expect that `@BeforeEach` and `@AfterEach` are executed around each test, dynamic or not:

    before
    Hello
    after
    before
    World
    after

## Steps to reproduce

```
import java.util.stream.Stream;
import org.junit.jupiter.api.*;
public class DynamicTestTest {

	@BeforeEach
	void beforeEach() {
		System.out.println("before");
	}
	@AfterEach
	void afterEach() {
		System.out.println("after");
	}

	@TestFactory
	Stream<DynamicTest> tests() {
		return Stream.of("Hello", "World")
				.map(word -> DynamicTest.dynamicTest(word, () -> System.out.println(word)));
	}

}
```

## Context

 - Used versions (Jupiter/Vintage/Platform): Jupiter 5.7.0-M1, Platform 1.7.0-M1
 - Build Tool/IDE: Maven (3.6.1), IntelliJ (2020.1.3)

**Comment by ogregoire** [comment]
Well, thank you GitHub, I searched for the same issue before creating this one. At least Google helped, only after I created this issue. It can be closed as a duplicate of #378.

**Comment by marcphilipp** [comment]
@ogregoire No worries! 🙂
